File: rlm/server/openai_compat.py
"""
RLM OpenAI-Compatible API Layer — Fase 9.2

Implementa POST /v1/chat/completions compatível com a SDK OpenAI.
Qualquer cliente que fala com a API OpenAI (LangChain, Cursor, n8n,
litellm, scripts Python com `openai` SDK) passa a falar com o RLM
sem mudança de código — só troca a `base_url`.

Análogo ao openai-http.ts do OpenClaw Gateway.

Diferença chave vs OpenClaw:
- OpenClaw: repassa para um LLM real cada turno
- RLM: o "chat" é continuação de sessão multi-turno com memória vetorial,
  REPL Python, tools, e decay temporal. O histórico não é passado
  pela request — está na sessão RLM.

Mapeamento de campo:
    OpenAI `messages[-1].content` → prompt para o agente RLM
    OpenAI `model`                → sobrescreve RLM_MODEL (opcional)
    OpenAI `user`                 → client_id (sessão RLM)
    OpenAI `stream`               → True = SSE, False = JSON completo
    OpenAI `max_tokens`           → ignorado (RLM tem max_iterations)
    OpenAI `temperature`          → ignorado (LLM configurado no RLM)

Autenticação:
    Authorization: Bearer {RLM_API_TOKEN}
    RLM_API_TOKEN não configurado → endpoint desabilitado.

Exemplo Python:
    from openai import OpenAI
    client = OpenAI(
        base_url="http://localhost:5000/v1",
        api_key="meu_token",   # = RLM_API_TOKEN
    )
    resp = client.chat.completions.create(
        model="rlm",
        messages=[{"role": "user", "content": "liste as vendas de hoje"}],
    )
    print(resp.choices[0].message.content)

Exemplo stream:
    for chunk in client.chat.completions.create(
        model="rlm",
        messages=[{"role": "user", "content": "faça uma análise longa"}],
        stream=True,
    ):
        print(chunk.choices[0].delta.content or "", end="", flush=True)
"""
from __future__ import annotations

import json
import os
import time
import uuid
from typing import Any, Iterator

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel, Field

from rlm.core.structured_log import get_logger
from rlm.gateway.auth_helpers import token_matches, authenticate_request
from rlm.server.runtime_pipeline import build_runtime_dispatch_services, dispatch_runtime_prompt_sync

compat_log = get_logger("openai_compat")

RLM_MODEL_ID = "rlm"  # model name reportado ao cliente


# ---------------------------------------------------------------------------
# Schemas OpenAI
# ---------------------------------------------------------------------------

class ChatMessage(BaseModel):
    role: str
    content: str | list | None = None
    name: str | None = None


class ChatCompletionRequest(BaseModel):
    model: str = RLM_MODEL_ID
    messages: list[ChatMessage]
    stream: bool = False
    user: str | None = None          # → client_id no RLM
    max_tokens: int | None = None    # ignorado
    temperature: float | None = None # ignorado
    n: int = 1                       # apenas 1 choice


# ---------------------------------------------------------------------------
# Builders de resposta no formato OpenAI
# ---------------------------------------------------------------------------

def _make_completion_response(
    run_id: str,
    model: str,
    content: str,
    finish_reason: str = "stop",
) -> dict:
    """Resposta JSON completa (stream=False)."""
    return {
        "id": f"chatcmpl-{run_id}",
        "object": "chat.completion",
        "created": int(time.time()),
        "model": model,
        "choices": [
            {
                "index": 0,
                "message": {"role": "assistant", "content": content},
                "finish_reason": finish_reason,
            }
        ],
        "usage": {
            # RLM não contabiliza tokens nativo — retorna -1 como placeholder
            "prompt_tokens": -1,
            "completion_tokens": -1,
            "total_tokens": -1,
        },
    }


def _iter_sse_chunks(
    run_id: str,
    model: str,
    content: str,
    chunk_size: int = 20,
) -> Iterator[str]:
    """
    Gera events SSE para simular streaming.

    O RLM roda o agente de forma síncrona (uma única execução),
    então o conteúdo já está completo ao retornar. Emitimos em
    chunks para compatibilidade com clientes que esperam SSE.
    """
    created = int(time.time())

    # Role chunk
    role_chunk = {
        "id": f"chatcmpl-{run_id}",
        "object": "chat.completion.chunk",
        "created": created,
        "model": model,
        "choices": [{"index": 0, "delta": {"role": "assistant"}, "finish_reason": None}],
    }
    yield f"data: {json.dumps(role_chunk)}\n\n"

    # Content chunks
    for i in range(0, len(content), chunk_size):
        piece = content[i : i + chunk_size]
        content_chunk = {
            "id": f"chatcmpl-{run_id}",
            "object": "chat.completion.chunk",
            "created": created,
            "model": model,
            "choices": [{"index": 0, "delta": {"content": piece}, "finish_reason": None}],
        }
        yield f"data: {json.dumps(content_chunk)}\n\n"

    # Stop chunk
    stop_chunk = {
        "id": f"chatcmpl-{run_id}",
        "object": "chat.completion.chunk",
        "created": created,
        "model": model,
        "choices": [{"index": 0, "delta": {}, "finish_reason": "stop"}],
    }
    yield f"data: {json.dumps(stop_chunk)}\n\n"
    yield "data: [DONE]\n\n"


# ---------------------------------------------------------------------------
# Router Factory
# ---------------------------------------------------------------------------

def create_openai_compat_router(expected_token: str) -> APIRouter:
    """
    Cria e retorna um APIRouter com o endpoint /v1/chat/completions.

    Args:
        expected_token: Token Bearer para autenticação.
                        Vazio → sem autenticação (inseguro, só para dev).

    Usage em api.py:
        from rlm.server.openai_compat import create_openai_compat_router
        token = os.environ.get("RLM_API_TOKEN", "")
        app.include_router(create_openai_compat_router(token), prefix="")
    """
    router = APIRouter(tags=["OpenAI-Compatible API"])

    def _auth(request: Request):
        """Authenticate via clients table → legacy token fallback.
        Returns ClientIdentity or None.

        Quando nenhum token está configurado (nem expected_token, nem env),
        retorna 503 — endpoint desabilitado por design (sem auth = inseguro).
        """
        from rlm.gateway.auth_helpers import configured_tokens as _ct

        sm = getattr(request.app.state, "session_manager", None)
        _raw = getattr(sm, "db_path", None) if sm else None
        db_path = _raw if isinstance(_raw, str) else "rlm_sessions.db"

        has_any_token = bool(expected_token) or bool(_ct("RLM_API_TOKEN"))
        if not has_any_token:
            raise HTTPException(503, "api authentication is not configured")

        return authenticate_request(
            request,
            db_path=db_path,
            env_names=("RLM_API_TOKEN",),
            scope="api",
            require=True,
            extra_legacy_tokens=(expected_token,) if expected_token else (),
        )

    def _extract_prompt(body: ChatCompletionRequest) -> str:
        """
        Extrai o prompt do último user message.
        Suporta content como string ou lista de partes (multimodal OpenAI).
        """
        for msg in reversed(body.messages):
            if msg.role == "user":
                content = msg.content
                if isinstance(content, str):
                    return content.strip()
                if isinstance(content, list):
                    # multimodal — concatena partes de texto
                    parts = []
                    for part in content:
                        if isinstance(part, dict) and part.get("type") == "text":
                            parts.append(str(part.get("text", "")))
                    return " ".join(parts).strip()
        # system prompt only?
        for msg in body.messages:
            if msg.role == "system" and isinstance(msg.content, str):
                return msg.content.strip()
        return ""

    @router.post("/v1/chat/completions")
    async def chat_completions(body: ChatCompletionRequest, request: Request):
        """
        OpenAI-compatible chat completion endpoint.

        Mapeia a última mensagem `user` como prompt para o agente RLM.
        Suporta stream=True (SSE) e stream=False (JSON).
        """
        identity = _auth(request)

        prompt = _extract_prompt(body)
        if not prompt:
            raise HTTPException(422, "No user message found in 'messages'")

        # client_id: identity > body.user > header > UUID
        client_id = (
            (identity.client_id if identity and identity.client_id != "legacy" else "")
            or body.user
            or request.headers.get("X-Session-ID", "")
            or f"openai_compat_{uuid.uuid4().hex[:8]}"
        ).strip()

        # model reportado ao cliente
        model = body.model if body.model != "gpt" else RLM_MODEL_ID

        run_id = uuid.uuid4().hex[:12]

        compat_log.info(
            f"[openai_compat] client_id={client_id!r} model={model!r} "
            f"stream={body.stream} prompt={prompt[:60]!r}"
        )

        # --- Executar via supervisor ou runtime pipeline do daemon ---
        sm = request.app.state.session_manager
        supervisor = request.app.state.supervisor
        hooks = getattr(request.app.state, "hooks", None)
        recursion_daemon = getattr(request.app.state, "recursion_daemon", None)

        session = sm.get_or_create(client_id)

        # ── Camada 3: populate session metadata from client identity ──
        if identity and identity.client_id != "legacy":
            if identity.preferred_channel:
                session.metadata["preferred_channel"] = identity.preferred_channel
            if identity.broadcast_channels:
                session.metadata["broadcast_channels"] = identity.broadcast_channels
            if identity.context_hint:
                session.metadata["context_hint"] = identity.context_hint
            session.metadata["client_profile"] = identity.profile

        if hooks:
            hooks.trigger(
                "openai_compat.request",
                session_id=session.session_id,
                client_id=client_id,
                model=model,
            )

        import asyncio
        loop = asyncio.get_event_loop()
        try:
            if recursion_daemon is not None:
                services = build_runtime_dispatch_services(request.app.state)
                dispatch_result = await loop.run_in_executor(
                    None,
                    lambda: dispatch_runtime_prompt_sync(
                        services,
                        client_id,
                        {
                            "text": prompt,
                            "content_type": "text",
                            "metadata": {
                                "model": model,
                                "transport": "openai_compat",
                                "run_id": run_id,
                            },
                        },
                        session=session,
                        source_name="api",
                    ),
                )
                content = str(dispatch_result.get("response", "") or "")
                finish_reason = "stop" if dispatch_result.get("status") != "error" else "error"
            else:
                result = await loop.run_in_executor(
                    None,
                    lambda: supervisor.execute(session, prompt),
                )
                content = result.response or ""
                finish_reason = "stop" if result.status != "error" else "error"
        except Exception as e:
            compat_log.warn(f"[openai_compat] Execution failed: {e}")
            raise HTTPException(500, f"Agent execution failed: {e}")

        sm.update_session(session)

        if body.stream:
            return StreamingResponse(
                _iter_sse_chunks(run_id, model, content),
                media_type="text/event-stream",
                headers={
                    "Cache-Control": "no-cache",
                    "X-Accel-Buffering": "no",
                    "X-Session-ID": session.session_id,
                },
            )

        return JSONResponse(
            content=_make_completion_response(run_id, model, content, finish_reason),
            headers={"X-Session-ID": session.session_id},
        )

    @router.get("/v1/models")
    async def list_models(request: Request):
        """
        Lista de modelos disponíveis (compatível com openai.models.list()).
        Retorna apenas o modelo RLM para identificação.
        """
        _auth(request)  # validate auth (identity not needed here)
        return {
            "object": "list",
            "data": [
                {
                    "id": RLM_MODEL_ID,
                    "object": "model",
                    "created": 1700000000,
                    "owned_by": "rlm",
                    "permission": [],
                }
            ],
        }

    return router

File: packages/gateway/src/openai-compat.ts
/**
 * OpenAI-Compatible API Layer — POST /v1/chat/completions
 *
 * Analogia Python: rlm/server/openai_compat.py
 *
 * Qualquer cliente que usa a API OpenAI (LangChain, Cursor, n8n, litellm,
 * openai SDK) fala com o RLM Brain sem mudança de código — só troca a base_url.
 *
 * Diferença chave:
 *   - OpenAI real: repassa para um LLM a cada turno
 *   - RLM: o "chat" é continuação de sessão multi-turno com memória vetorial,
 *     tools, e decay temporal. O histórico não é passado pela request — está
 *     na sessão RLM. Apenas messages[-1].content é usado como prompt.
 *
 * Mapeamento de campos:
 *   OpenAI messages[-1].content → prompt para o brain
 *   OpenAI model                → sobrescreve RLM_MODEL (opcional, ignorado)
 *   OpenAI user                 → client_id (sessão RLM)
 *   OpenAI stream               → true = SSE chunked, false = JSON completo
 *   OpenAI max_tokens           → ignorado (RLM tem max_iterations)
 *   OpenAI temperature          → ignorado (configurado no Brain)
 *
 * Autenticação:
 *   Authorization: Bearer {RLM_API_TOKEN}
 *   Endpoint desabilitado se RLM_API_TOKEN não configurado.
 *
 * Exemplo JS:
 *   import OpenAI from "openai";
 *   const client = new OpenAI({ baseURL: "http://localhost:3000/v1", apiKey: "token" });
 *   const res = await client.chat.completions.create({
 *     model: "rlm",
 *     messages: [{ role: "user", content: "liste as vendas de hoje" }],
 *   });
 */

import type { Context } from "hono";
import { Hono } from "hono";
import { stream } from "hono/streaming";
import { z } from "zod";
import { childLogger } from "./logger.js";
import { newEnvelope } from "./envelope.js";
import type { ChannelRegistry } from "./registry.js";
import type { WsBridge } from "./ws-bridge.js";
import type { Envelope } from "./envelope.js";

const log = childLogger({ component: "openai-compat" });

const RLM_MODEL_ID = "rlm";

// ---------------------------------------------------------------------------
// Schemas OpenAI (subconjunto mínimo)
// ---------------------------------------------------------------------------

const ChatMessageSchema = z.object({
  role: z.enum(["system", "user", "assistant", "tool"]),
  content: z.union([z.string(), z.array(z.unknown())]).nullable().default(""),
  name: z.string().optional(),
});

const ChatCompletionRequestSchema = z.object({
  model: z.string().default(RLM_MODEL_ID),
  messages: z.array(ChatMessageSchema).min(1),
  stream: z.boolean().default(false),
  user: z.string().optional(),       // → client_id no RLM
  max_tokens: z.number().optional(), // ignorado
  temperature: z.number().optional(), // ignorado
  n: z.number().default(1),
});

type ChatCompletionRequest = z.infer<typeof ChatCompletionRequestSchema>;

// ---------------------------------------------------------------------------
// Builders de resposta no formato OpenAI
// ---------------------------------------------------------------------------

function makeCompletionResponse(
  runId: string,
  model: string,
  content: string,
  finishReason = "stop",
): object {
  return {
    id: `chatcmpl-${runId}`,
    object: "chat.completion",
    created: Math.floor(Date.now() / 1000),
    model,
    choices: [
      {
        index: 0,
        message: { role: "assistant", content },
        finish_reason: finishReason,
      },
    ],
    // RLM não contabiliza tokens — retorna -1 como placeholder
    usage: { prompt_tokens: -1, completion_tokens: -1, total_tokens: -1 },
  };
}

function makeChunkResponse(
  runId: string,
  model: string,
  delta: string,
  finishReason: string | null = null,
): object {
  return {
    id: `chatcmpl-${runId}`,
    object: "chat.completion.chunk",
    created: Math.floor(Date.now() / 1000),
    model,
    choices: [
      {
        index: 0,
        delta: finishReason ? {} : { role: "assistant", content: delta },
        finish_reason: finishReason,
      },
    ],
  };
}

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

function validateApiToken(c: Context, expected: string): boolean {
  const auth = c.req.header("authorization");
  if (!auth?.startsWith("Bearer ")) return false;
  const provided = auth.slice(7);
  // Timing-safe
  const a = Buffer.from(provided.padEnd(256, "\0"));
  const b = Buffer.from(expected.padEnd(256, "\0"));
  return a.length === b.length && a.equals(b);
}

// ---------------------------------------------------------------------------
// Reply collector — espera resposta do Brain via WsBridge
// ---------------------------------------------------------------------------

function waitForBrainReply(
  bridge: WsBridge,
  correlationId: string,
  timeoutMs = 60_000,
): Promise<string> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      unsub();
      reject(new Error("Brain reply timeout"));
    }, timeoutMs);

    const unsub = bridge.onReply((envelope: Envelope) => {
      if (
        envelope.correlation_id === correlationId ||
        envelope.id === correlationId
      ) {
        clearTimeout(timer);
        unsub();
        resolve(envelope.text ?? "");
      }
    });
  });
}

// ---------------------------------------------------------------------------
// Router factory
// ---------------------------------------------------------------------------

export interface OpenAICompatConfig {
  /** Lê RLM_API_TOKEN se não fornecido. Desabilitado se vazio. */
  apiToken?: string;
  /** Timeout de resposta do Brain em ms (default: 60s) */
  brainTimeoutMs?: number;
}

export function createOpenAICompatRouter(
  registry: ChannelRegistry,
  bridge: WsBridge,
  config: OpenAICompatConfig = {},
): Hono {
  const apiToken = config.apiToken ?? process.env["RLM_API_TOKEN"] ?? "";
  const brainTimeout = config.brainTimeoutMs ?? 60_000;

  const router = new Hono();

  // Models list (compatibilidade)
  router.get("/v1/models", (c) => {
    if (apiToken && !validateApiToken(c, apiToken)) {
      return c.json({ error: { message: "Unauthorized", type: "authentication_error" } }, 401);
    }
    return c.json({
      object: "list",
      data: [{ id: RLM_MODEL_ID, object: "model", created: 0, owned_by: "rlm" }],
    });
  });

  // Chat completions
  router.post("/v1/chat/completions", async (c) => {
    // Endpoint desabilitado se token não configurado
    if (!apiToken) {
      return c.json(
        { error: { message: "OpenAI compat endpoint not configured", type: "invalid_request_error" } },
        404,
      );
    }

    if (!validateApiToken(c, apiToken)) {
      return c.json(
        { error: { message: "Invalid API key", type: "authentication_error" } },
        401,
      );
    }

    let req: ChatCompletionRequest;
    try {
      const raw = await c.req.json() as unknown;
      req = ChatCompletionRequestSchema.parse(raw);
    } catch {
      return c.json(
        { error: { message: "Invalid request body", type: "invalid_request_error" } },
        400,
      );
    }

    // Extrai último conteúdo de usuário como prompt
    const lastUserMsg = [...req.messages].reverse().find((m) => m.role === "user");
    const prompt = typeof lastUserMsg?.content === "string"
      ? lastUserMsg.content
      : JSON.stringify(lastUserMsg?.content ?? "");

    if (!prompt.trim()) {
      return c.json(
        { error: { message: "No user message found", type: "invalid_request_error" } },
        400,
      );
    }

    const clientId = req.user ?? "openai_compat";
    const runId = crypto.randomUUID().replace(/-/g, "");

    const envelope = newEnvelope({
      source_channel: "api",
      source_id: clientId,
      source_client_id: `api:${clientId}`,
      direction: "inbound",
      message_type: "text",
      text: prompt,
    });

    // Reusa o id do envelope como correlation_id para rastrear a resposta
    const correlationId = envelope.id;
    const forwarded = registry.forwardToBrain(envelope);
    if (!forwarded) {
      return c.json(
        { error: { message: "Brain unavailable", type: "service_unavailable" } },
        503,
      );
    }

    log.info({ clientId, promptLen: prompt.length, stream: req.stream }, "OpenAI compat request");

    if (req.stream) {
      // SSE streaming
      return stream(c, async (writer) => {
        try {
          const text = await waitForBrainReply(bridge, correlationId, brainTimeout);

          // Emite o conteúdo completo como um único chunk (Brain não faz streaming ainda)
          const chunk = makeChunkResponse(runId, req.model, text);
          await writer.write(`data: ${JSON.stringify(chunk)}\n\n`);

          // Chunk final
          const done = makeChunkResponse(runId, req.model, "", "stop");
          await writer.write(`data: ${JSON.stringify(done)}\n\n`);
          await writer.write("data: [DONE]\n\n");
        } catch (err) {
          log.error({ err, clientId }, "OpenAI compat stream error");
          const errChunk = { error: { message: String(err), type: "internal_server_error" } };
          await writer.write(`data: ${JSON.stringify(errChunk)}\n\n`);
        }
      });
    }

    // Resposta completa (non-stream)
    try {
      const text = await waitForBrainReply(bridge, correlationId, brainTimeout);
      return c.json(makeCompletionResponse(runId, req.model, text));
    } catch (err) {
      log.error({ err, clientId }, "OpenAI compat timeout/error");
      return c.json(
        { error: { message: "Brain response timeout", type: "internal_server_error" } },
        504,
      );
    }
  });

  return router;
}
